sdk/python/kfp/components/type_annotation_utils.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2021 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for handling Python type annotation."""
+
+import re
+from typing import TypeVar, Union
+
+T = TypeVar('T')
+
+
+def extract_type_with_optional(annotation: T) -> T:
+  """Extracts the type from Optional[<type>] if applicable.
+
+  For example::
+    Optional[str] -> str
+    str -> str
+    List[int] -> List[int]
+
+  Args:
+    annotation: The original type annotation which may or may not has
+      `Optional`.
+
+  Returns:
+    The type inside Optional[] if Optional exists, otherwise the original type.
+  """
+  if getattr(annotation, '__origin__',
+             None) is Union and annotation.__args__[1] is type(None):
+    return annotation.__args__[0]
+  return annotation
+
+
+def get_short_type_name(type_name: str) -> str:
```

Comment:
Why can't we introspect the actual type instead of doing string matching here? 

We may be able to use `typing_extensions` to do this reliably.

Reply:
Python's generic "types" are not really types. The information is semi-erased and moved to internal fields that change with versions. This is possible to do (I do this when converting annotations to type structures), but it's not trivial.

Reply:
> Why can't we introspect the actual type instead of doing string matching here?

yes, we should probably revisit this function. But at this moment, I believe it's called only at places where an actual type has been converted into a string already.